### PR TITLE
New autodetection method KubernetesInternalIP added

### PR DESCRIPTION
**Pull request by hanamantagoudvk**
## Description

New autodetection method KubernetesInternalIP / KubernetesInternalIP6 being introduced so that calico can always pick the node IP deterministically. 

Fixes https://github.com/projectcalico/calico/issues/4870

Fixes https://github.com/projectcalico/calico/issues/5090

### CODE REVIEW

**Comment by caseydavenport** [comment]
Tagging for v3.22

File: pkg/lifecycle/startup/startup.go [hunk 1]
```diff
@@ -772,6 +777,51 @@ func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int) *cnet.IPN
 	return cidr
 }
 
+// autoDetectUsingK8sInternalIP reads K8s Node InternalIP.
+func autoDetectUsingK8sInternalIP(version int, nodeName string) *cnet.IPNet {
+	// If running under kubernetes with secrets to call k8s API
+	if config, err := rest.InClusterConfig(); err == nil {
```
Comment (tomastigera) [inline pkg/lifecycle/startup/startup.go:783]: nit: flip the condition and return nil on error rather then indenting everything below?

File: pkg/lifecycle/startup/startup.go [hunk 2]
```diff
@@ -312,10 +320,32 @@ func MonitorIPAddressSubnets() {
 
 	pollInterval := getMonitorPollInterval()
 
+	var clientset *kubernetes.Clientset
+	var config *rest.Config
+	var k8sNode *v1.Node
+	var err error
+	if config, err = rest.InClusterConfig(); err != nil {
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/startup.go:327]: This code needs to work on non-kubernetes clusters as well - I'd suggest modeling after the logic above: https://github.com/projectcalico/node/pull/1242/files#diff-e481557bd1e2fed4cea80458ebb02e18797fdf49c42944f4a8d6c23cbfbec3f9R130

File: pkg/lifecycle/startup/startup.go [hunk 3]
```diff
@@ -772,6 +809,27 @@ func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int) *cnet.IPN
 	return cidr
 }
 
+// autoDetectUsingK8sInternalIP reads K8s Node InternalIP.
+func autoDetectUsingK8sInternalIP(version int, k8sNode *v1.Node) *cnet.IPNet {
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/startup.go:813]: should probably put this in the autodetection package along with the other methods.

Comment (caseydavenport) [reply pkg/lifecycle/startup/startup.go:813]: It would require moving the getLocalCIDR function with it, but I think that's OK

File: pkg/lifecycle/startup/startup.go [hunk 4]
```diff
@@ -772,6 +809,27 @@ func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int) *cnet.IPN
...
+	var address string
+	var err error
+
+	nodeAddresses := k8sNode.Status.Addresses
+	for _, addr := range nodeAddresses {
+		if addr.Type == v1.NodeInternalIP {
+			if (version == 4 && utils.IsIPv4String(addr.Address)) || (version == 6 && utils.IsIPv6String(addr.Address)) {
+				address, err = getLocalCIDR(addr.Address, version, autodetection.GetInterfaces)
+			}
+			if err != nil {
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/startup.go:823]: This error check should be within the branch above, just below `getLocalCIDR` on line 821.

File: pkg/lifecycle/startup/startup.go [hunk 5]
```diff
@@ -772,6 +809,27 @@ func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int) *cnet.IPN
...
+				return nil
+			}
+		}
+	}
+
+	ipNet := cnet.MustParseCIDR(address)
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/startup.go:829]: We should use `cnet.ParseCIDR` instead of `MustParseCIDR` - the latter is just for testing, and can result in panics. We want to handle any error gracefully.

**Review by caseydavenport — CHANGES_REQUESTED** [review]

File: pkg/lifecycle/startup/startup.go [hunk 6]
```diff
@@ -772,6 +777,53 @@ func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int) *cnet.IPN
 	return cidr
 }
 
+// autoDetectUsingK8sInternalIP reads K8s Node InternalIP.
+func autoDetectUsingK8sInternalIP(version int, nodeName string) *cnet.IPNet {
+	// If running under kubernetes with secrets to call k8s API
+	var config *rest.Config
+	var err error
+	if config, err = rest.InClusterConfig(); err != nil {
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply pkg/lifecycle/startup/startup.go]: huh, it's a little confusing - this is the make target: https://github.com/projectcalico/node/blob/5407871b224027b8d3d4e92ffa9a741a72c3facd/Makefile#L326-L334

So, `make fv` will run the tests (as well as some others)

**Comment by caseydavenport** [comment]
@hanamantagoudvk just checking in on this one - did you see my latest comments?

**Comment by hanamantagoudvk** [comment]
> @hanamantagoudvk just checking in on this one - did you see my latest comments?

@caseydavenport : Sorry for the delay. Uploaded the new patchset. I am still not able to run UT cases in my setup.
I get some error related to setup. I am working on UT part.

File: pkg/lifecycle/startup/startup.go (hunk 2 above)
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup.go]: modified as per suggestion

File: pkg/lifecycle/startup/startup.go (hunk 4 above)
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup.go]: done

File: pkg/lifecycle/startup/startup.go (hunk 5 above)
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup.go]: modified as per suggestion

File: pkg/lifecycle/startup/startup.go [hunk 7]
```diff
@@ -467,7 +469,7 @@ func configureIPsAndSubnets(node *libapi.Node) (bool, error) {
 	ipv4Env := os.Getenv("IP")
 	if ipv4Env == "autodetect" || (ipv4Env == "" && node.Spec.BGP.IPv4Address == "") {
 		adm := os.Getenv("IP_AUTODETECTION_METHOD")
-		cidr := autoDetectCIDR(adm, 4)
+		cidr := autoDetectCIDR(adm, 4, node.Name)
```
[the remark this answers is not in the document]
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup.go]: done

**Review by tomastigera — COMMENTED** [review]
I have nothing major against this change, tbh I like that it is coming, we did consider it for eBPF and then we went different way. So great! :+1:  I will leave it to @caseydavenport to approve.

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+	_, ipNet, err := cnet.ParseCIDR(address)
+	if err != nil {
+		log.Errorf("Unable to parse CIDR %v : %v", address, err)
+		return nil
+	}
+	return ipNet
+}
+
+// getLocalCIDR attempts to merge CIDR information from the host with the given IP address.
```
[the remark this answers is not in the document]
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/autodetection/autodetection_methods.go:189]: GetLocalCIDR is required in `startup_test.go`

File: pkg/lifecycle/startup/startup_test.go [hunk 9]
```diff
@@ -873,26 +924,45 @@ var _ = Describe("FV tests against a real etcd", func() {
...
 	DescribeTable("Test variations on how IPs are detected.",
-		func(node *libapi.Node, items []EnvItem, expected bool) {
+		func(node *libapi.Node, k8sNode *v1.Node, items []EnvItem, expected bool) {
 
 			for _, item := range items {
 				os.Setenv(item.key, item.value)
 			}
 
-			check, err := configureIPsAndSubnets(node)
+			check, err := configureIPsAndSubnets(node, k8sNode, getDefaultInterfacesToExclude())
```
[the remark this answers is not in the document]
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup_test.go:933]: yes modified

File: pkg/lifecycle/startup/startup_test.go [hunk 10]
```diff
@@ -78,6 +100,32 @@ func makeNode(ipv4 string, ipv6 string) *libapi.Node {
...
+	ip4net.IP = ip4.IP
+
+	ip6Addr := ""
+	if ipv6 != "" {
+		ip6, ip6net, _ := net.ParseCIDR(ipv6)
+		// Guard against nil here in case we pass in an empty string for IPv6.
+		if ip6 != nil {
+			ip6net.IP = ip6.IP
+		}
+		ip6Addr = ip6net.String()
```
[the remark this answers is not in the document]
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup_test.go:115]: yes i want to add various combinations with `IPv4 `and `IPv6`  addresses, but not able to mock the `GetInterfaces` method inside `autoDetectUsingK8sInternalIP`. Due to that , `configureIPsAndSubnets `always returns `false`

File: pkg/lifecycle/startup/startup_test.go [hunk 11]
```diff
@@ -78,6 +100,32 @@ func makeNode(ipv4 string, ipv6 string) *libapi.Node {
 	return n
 }
 
+// makeK8sNode creates an v1.Node with some Addresses populated.
+func makeK8sNode(ipv4 string, ipv6 string) *v1.Node {
```
[the remark this answers is not in the document]
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup_test.go:104]: same comment as below

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+	nodeAddresses := k8sNode.Status.Addresses
+	for _, addr := range nodeAddresses {
+		if addr.Type == v1.NodeInternalIP {
+			if (version == 4 && utils.IsIPv4String(addr.Address)) || (version == 6 && utils.IsIPv6String(addr.Address)) {
+				address, err = GetLocalCIDR(addr.Address, version, GetInterfaces)
+				if err != nil {
+					return nil
+				}
+			}
```
[the remark this answers is not in the document]
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/autodetection/autodetection_methods.go:177]: @caseydavenport , any inputs here ?

[the remark this answers is not in the document]
Comment (caseydavenport) [reply pkg/lifecycle/startup/autodetection/autodetection_methods.go:177]: Yeah I think it makes sense to pick the first address that we find in the list. It's a more expected behavior.

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go [hunk 13]
```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply pkg/lifecycle/startup/autodetection/autodetection_methods.go:1]: yeah, agreed.

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+		return nil
+	} else {
+		log.Infof("Using autodetected IPv%d address %s, detected by connecting to %s", version, cidr.String(), dest)
+		return cidr
+	}
+}
+
+// autoDetectCIDRBySkipInterface auto-detects the first valid Network on the interfaces
+// matching the supplied interface regexes.
+func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int, defInterfacesToExclude []string) *cnet.IPNet {
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/autodetection/autodetection_methods.go:149]: This didn't take interfaces to exclude as an argument before - why do we need this change?

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+	// The autodetection method is not recognised and is required.  Exit.
+	log.Errorf("Invalid IP autodetection method: %s", method)
+	utils.Terminate()
+	return nil
+}
+
+// autoDetectCIDRFirstFound auto-detects the first valid Network it finds across
+// all interfaces (excluding common known internal interface names).
+func autoDetectCIDRFirstFound(version int, defInterfacesToExclude []string) *cnet.IPNet {
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/autodetection/autodetection_methods.go:93]: This function takes a new argument now - is this really needed for this PR?

File: pkg/lifecycle/startup/startup.go [hunk 16]
```diff
@@ -170,9 +173,15 @@ func Run() {
...
+
+		k8sNode, err = clientset.CoreV1().Nodes().Get(ctx, k8sNodeName, metav1.GetOptions{})
+		if err != nil {
+			log.WithError(err).Error("Failed to read Node from datastore")
+			utils.Terminate()
+		}
 	}
 
-	configureAndCheckIPAddressSubnets(ctx, cli, node)
+	configureAndCheckIPAddressSubnets(ctx, cli, node, k8sNode, DEFAULT_INTERFACES_TO_EXCLUDE)
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/startup.go:184]: I think we should revert this change to pass DEFAULT_INTERFACES_TO_EXCLUDE all the way through the call stack of the autodetection functions. It seems to be an extra argument just to pass a constant value.

File: pkg/lifecycle/startup/startup_test.go [hunk 17]
```diff
@@ -52,6 +53,27 @@ func fakeExitFunction(ec int) {
 	exitCode = ec
 }
 
+func getDefaultInterfacesToExclude() []string {
```
Comment (caseydavenport) [inline pkg/lifecycle/startup/startup_test.go:56]: I think this function can be removed if we revert the change to pass the interfaces as an argument.

**Review by caseydavenport — CHANGES_REQUESTED** [review]
I mostly don't understand why we're now passing DEFAULT_INTERFACES_TO_EXCLUDE as an argument through several layers of call stack, especially since it's not used in the new autodetection method. We should revert that. 

Otherwise this is looking good.

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go (hunk 14 above)
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/autodetection/autodetection_methods.go]: yes removed it.

File: pkg/lifecycle/startup/autodetection/autodetection_methods.go (hunk 15 above)
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/autodetection/autodetection_methods.go]: yes removed it.

File: pkg/lifecycle/startup/startup_test.go (hunk 17 above)
Comment (hanamantagoudvk) [reply pkg/lifecycle/startup/startup_test.go]: yes removed it.

**Comment by caseydavenport** [comment]
@hanamantagoudvk you've got an import cycle looks like:

```
package command-line-arguments01:39
	imports github.com/projectcalico/node/pkg/lifecycle/startup01:39
	imports github.com/projectcalico/node/pkg/lifecycle/startup/autodetection01:39
	imports github.com/projectcalico/node/pkg/lifecycle/startup: import cycle not allowed01:39
Makefile:281: recipe for target 'dist/bin//calico-node-amd64' failed
```

**Comment by hanamantagoudvk** [comment]
```
package command-line-arguments01:39
	imports github.com/projectcalico/node/pkg/lifecycle/startup01:39
	imports github.com/projectcalico/node/pkg/lifecycle/startup/autodetection01:39
	imports github.com/projectcalico/node/pkg/lifecycle/startup: import cycle not allowed01:39
Makefile:281: recipe for target 'dist/bin//calico-node-amd64' failed
```

@caseydavenport 
The reason for above import cycle is `DEFAULT_INTERFACES_TO_EXCLUDE` . Due to split in code now , `DEFAULT_INTERFACES_TO_EXCLUDE` has remained in `startup` package and its getting used in `autodetection `package. `autodetection `package imports `startup `package and vice-versa.

**Comment by caseydavenport** [comment]
> The reason for above import cycle is DEFAULT_INTERFACES_TO_EXCLUDE . Due to split in code now , DEFAULT_INTERFACES_TO_EXCLUDE has remained in startup package and its getting used in autodetection package. autodetection package imports startup package and vice-versa.

Should probably just move DEFAULT_INTERFACES_TO_EXCLUDE to the autodetection package then, right?

**Comment by hanamantagoudvk** [comment]
> > The reason for above import cycle is DEFAULT_INTERFACES_TO_EXCLUDE . Due to split in code now , DEFAULT_INTERFACES_TO_EXCLUDE has remained in startup package and its getting used in autodetection package. autodetection package imports startup package and vice-versa.
> 
> Should probably just move DEFAULT_INTERFACES_TO_EXCLUDE to the autodetection package then, right?

@caseydavenport 
yes now i have moved DEFAULT_INTERFACES_TO_EXCLUDE into the autodetection pkg.

**Comment by hanamantagoudvk** [comment]
@caseydavenport : Please review it and let me know if anything is pending ? I am not able to see the above errors since i dont have access to CI system.

**Review by caseydavenport — APPROVED** [review]
This LGTM - Thanks for all of the hard work on this @hanamantagoudvk

**Comment by caseydavenport** [comment]
@hanamantagoudvk just a static check failure it looks like:

```
pkg/lifecycle/startup/startup_test.go:86:5: composites: `k8s.io/api/core/v1.NodeAddress` composite literal uses unkeyed fields (govet)04:51
				{v1.NodeInternalIP, ipv4},04:51
				^04:51
pkg/lifecycle/startup/startup_test.go:87:5: composites: `k8s.io/api/core/v1.NodeAddress` composite literal uses unkeyed fields (govet)04:52
				{v1.NodeInternalIP, ipv6},04:52
				^
```

**Comment by hanamantagoudvk** [comment]
@caseydavenport , what else is failing now ?

**Comment by caseydavenport** [comment]
I think they were test flakes, kicked off another build.

**Comment by caseydavenport** [comment]
```
UT for Node IP assignment and conflict checking. Test variations on how IPs are detected. [It] Test with no "IP" env var set 02:25
/go/src/github.com/projectcalico/node/pkg/lifecycle/startup/startup_test.go:90602:25
  Expected02:25
      <bool>: false02:25
  to equal02:25
      <bool>: true02:25
  /go/src/github.com/projectcalico/node/pkg/lifecycle/startup/startup_test.go:902
```

This test appears to be failing consistently

**Comment by caseydavenport** [comment]
@hanamantagoudvk thanks for sticking with this! Glad we got it merged.

What was the last issue with that test? I couldn't tell because the commits were squashed.

Also, are you planning to do documentation for this PR? If so I can help.

Similarly, we probably want operator support for this feature as well - I can pick that up if you're strapped for time.

**Comment by hanamantagoudvk** [comment]
@caseydavenport , basically ginkgo orders the testcases alphabetically. So the tests i added were run before the existing one and it was setting the `IP` env variable and not resetting it. So due to that the testcases were failing. In my local setup , it used to pass on and off. So now i put a code to reset those env variables.

Regarding documentation , yes i can do the documentation. 

Ya that operator support , you can pick that up.

**Comment by hanamantagoudvk** [comment]
@caseydavenport , regarding documentation , i believe in calico website where different autodetection methods are listed , we need to this new one as well. Other than anywhere else we need to capture this one ?

If i need to modify the calico website content , which repo i need to look at and modify it ?
